src/boh_app/data/process_autosave.py

```diff
@@ -118,3 +120,18 @@ def get_recipes(self, data: dict[str, Any]) -> list[KnownRecipe]:
                 else:
                     known_recipe_skills_mapping[recipe_id].append(SkillRef(id=skill_id))
         return [KnownRecipe(id=k, skills=v) for k, v in known_recipe_skills_mapping.items()]
+
+    def get_books(self, data: dict[str, Any]) -> list[KnownRecipe]:
+        root_data = data["RootPopulationCommand"]["Spheres"][3]
+        assert root_data["GoverningSphereSpec"]["Id"] == "Library"
+        known_books = set()
+        for lib_token in root_data["Tokens"]:
+            for dominion in lib_token["Payload"]["Dominions"]:
+                for sphere in dominion["Spheres"]:
+                    if sphere["GoverningSphereSpec"]["Label"] == "BOOKSHELF":
+                        for bookshelf_token in sphere["Tokens"]:
+                            payload = bookshelf_token["Payload"]
+                            if any(mut.startswith("mastery.") for mut in payload["Mutations"].keys()):
+                                known_books.add(payload["EntityId"])
```

Comment:
you can make this less nested by using a comprehension:

```suggestion
        books = [
            bookshelf_token["Payload"]
            for lib_token in root_data["Tokens"]
            for dominion in lib_token["Payload"]["Dominions"]
            for sphere in dominion["Spheres"]
            if sphere["GoverningSphereSpec"]["Label"] == "BOOKSHELF"
            for bookshelf_token in sphere["Tokens"]
        ]
        known_books = {
            book["EntityId"]
            for book in books
            if any(mut.startswith("mastery.") for mut in payload["Mutations"])
        }
```

Reply:
mastered_known, unread_known = partition(lambda book: any(mut.startswith("mastery.") for mut in book["Mutations"]), book)